default/python/AI/savegame_codec/_decoder.py

```diff
@@ -11,81 +11,41 @@
                            SET_PREFIX, TRUE, TUPLE_PREFIX, trusted_classes, )
 
 
-def _ensure_binary(s, encoding='utf-8', errors='strict'):
-    """Coerce **s** to bytes.
-      - `str` -> encoded to `bytes`
-      - `bytes` -> `bytes`
+class SaveDecompressException(Exception):
     """
-    if isinstance(s, str):
-        return s.encode(encoding, errors)
-    elif isinstance(s, bytes):
-        return s
-    else:
-        raise TypeError("not expecting type '%s'" % type(s))
-
-
-def _ensure_str(s, encoding='utf-8', errors='strict'):
-    """Coerce *s* to `str`.
-      - `str` -> `str`
-      - `bytes` -> decoded to `str`
+    Exception class for troubles with decompressing save game string.
     """
-    if not isinstance(s, (str, bytes)):
-        raise TypeError("not expecting type '%s'" % type(s))
-    elif isinstance(s, bytes):
-        s = s.decode(encoding, errors)
-    return s
+    pass
 
 
 @profile
 def load_savegame_string(string: str):
     """
     :rtype: AIstate
+    :raises: SaveUncompressException, InvalidSaveGameException
     """
     import base64
     import zlib
 
-    new_string = string
     try:
         new_string = base64.b64decode(string.encode("utf-8"))
-    except TypeError as e:
-        # The base64 module docs only mention a TypeError exception, for wrong padding
-        # Older save files won't be base64 encoded, but seemingly that doesn't trigger
-        # an exception here;
-        debug("When trying to base64 decode savestate got exception: %s" % e)
-    try:
-        # depending on a previous try new_string can be str or bytes
-        new_string = zlib.decompress(_ensure_binary(new_string))
-    except zlib.error:
-        pass  # probably an uncompressed (or wrongly base64 decompressed) string
-    try:
-        # depending on a previous try new_string can be str or bytes
-        decoded_state = decode(_ensure_str(new_string))
-        debug("Decoded a zlib-compressed and apparently base64-encoded save-state string.")
-        return decoded_state
-    except (InvalidSaveGameException, ValueError, TypeError) as e:
-        debug("Base64/zlib decoding path for savestate failed: %s" % e)
-
+    except binascii.Error as e:
+        raise SaveDecompressException("Fail to decode base64 savestate %s" % e) from e
     try:
-        string = zlib.decompress(string.encode('utf-8'))
-        debug("zlib-decompressed a non-base64-encoded save-state string.")
-    except zlib.error:
-        # probably an uncompressed string
-        debug("Will try decoding savestate string without base64 or zlib compression.")
-
-    # We need to check uncompressed string, since zip archive of the empty string is not the empty string.
-    if not string:
-        raise Exception("Save game string is empty")
-    return decode(string)
+        new_string = zlib.decompress(new_string)
+    except zlib.error as e:
+        raise SaveDecompressException("Fail to ungzip savestate %s" % e) from e
```

Comment:
typo: unzip

Reply:
zlib docs:
The functions in this module allow compression and decompression using the
zlib library, which is based on GNU zip.

So it should be `gunzip` like corresponding tool.   But I think I will use decompress to reduce confusion.